File: CPP/fundamentals/exceptionHandling.cpp
#include <iostream>
#include <exception>

void exceptionHandlingSummary() {
    //Basic try-catch
    try {
        int x = 10, y = 0;
        if (y == 0) {
            throw std::runtime_error("Division by zero");
        }
        int result = x / y;
    }
    catch (const std::runtime_error& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }
    catch (const std::exception& e) { //Catch base class
        std::cerr << "Exception: " << e.what() << std::endl;
    }
    catch (...) { //Catch all
        std::cerr << "Unknown exception" << std::endl;
    }
    
    //Multiple exception types
    try {
        std::vector<int> vec = {1, 2, 3};
        vec.at(10); //Throws out_of_range
    }
    catch (const std::out_of_range& e) {
        std::cerr << "Out of range: " << e.what() << std::endl;
    }
    
    //Common standard exceptions:
    //std::runtime_error - general runtime errors
    //std::logic_error - programming logic errors
    //std::out_of_range - access beyond bounds
    //std::invalid_argument - invalid function argument
    //std::bad_alloc - memory allocation failure
    
    //Custom exception
    class MyException : public std::exception {
    public:
        const char* what() const noexcept override {
            return "My custom error";
        }
    };
    
    try {
        throw MyException();
    }
    catch (const MyException& e) {
        std::cerr << e.what() << std::endl;
    }
    
    //Rethrowing
    try {
        try {
            throw std::runtime_error("Error");
        }
        catch (...) {
            std::cout << "Caught, now rethrowing" << std::endl;
            throw; //Rethrow same exception
        }
    }
    catch (const std::exception& e) {
        std::cerr << "Caught again: " << e.what() << std::endl;
    }
}

File: Java/fundamentals/ExceptionHandling.java
import java.io.*;

public class ExceptionHandlingReference {
    public static void main(String[] args) {
        tryCatchUsage();
        multipleCatchUsage();
        tryWithResourcesUsage();
        throwUsage();
        throwsUsage();
        customExceptionUsage();
    }

    //============================================================
    //TRY-CATCH - USAGE
    //============================================================
    public static void tryCatchUsage() {
        //Basic try-catch
        try {
            int result = 10 / 0; //ArithmeticException
        } catch(ArithmeticException e) {
            System.out.println("Cannot divide by zero");
        }
        
        //Access exception details
        try {
            String str = null;
            str.length(); //NullPointerException
        } catch(NullPointerException e) {
            System.out.println(e.getMessage());
            e.printStackTrace();
        }
        
        //Finally always executes (cleanup code)
        try {
            int result = 10 / 2;
            System.out.println(result);
        } catch(Exception e) {
            System.out.println("Exception occurred");
        } finally {
            System.out.println("Always executes"); //Cleanup, close resources
        }
        
        //Finally executes even with return
        try {
            return;
        } finally {
            System.out.println("Executes before return");
        }
    }

    //============================================================
    //MULTIPLE CATCH - USAGE
    //============================================================
    public static void multipleCatchUsage() {
        //Multiple catch blocks - order matters (specific to general)
        try {
            int[] arr = {1, 2, 3};
            System.out.println(arr[5]); //ArrayIndexOutOfBoundsException
        } catch(ArrayIndexOutOfBoundsException e) {
            System.out.println("Index out of bounds");
        } catch(Exception e) {
            System.out.println("General exception");
        }
        
        //Multi-catch (Java 7+) - same handling for multiple exceptions
        try {
            String str = null;
            str.length();
        } catch(NullPointerException | ArithmeticException e) {
            System.out.println("Null or arithmetic error");
        }
        
        //Nested try-catch
        try {
            try {
                int result = 10 / 0;
            } catch(ArithmeticException e) {
                System.out.println("Inner catch");
                throw e; //Re-throw to outer catch
            }
        } catch(ArithmeticException e) {
            System.out.println("Outer catch");
        }
    }

    //============================================================
    //TRY-WITH-RESOURCES - USAGE (Auto-close resources)
    //============================================================
    public static void tryWithResourcesUsage() {
        //Single resource
        try(BufferedReader br = new BufferedReader(new FileReader("file.txt"))) {
            String line = br.readLine();
        } catch(IOException e) {
            System.out.println("File error");
        }
        
        //Multiple resources
        try(FileReader fr = new FileReader("input.txt");
            BufferedReader br = new BufferedReader(fr);
            FileWriter fw = new FileWriter("output.txt")) {
            String line = br.readLine();
            fw.write(line);
        } catch(IOException e) {
            System.out.println("IO error");
        }
    }

    //============================================================
    //THROW - USAGE (Throw exception explicitly)
    //============================================================
    public static void throwUsage() {
        //Throw built-in exception
        int age = -5;
        if(age < 0) {
            throw new IllegalArgumentException("Age cannot be negative");
        }
        
        //Re-throw exception to be caught again later
        try {
            int result = 10 / 0;
        } catch(ArithmeticException e) {
            System.out.println("Logging error");
            throw e; //Re-throw to caller
        }
    }
    
    //Single custom throw, definition is class below that extends Exception
    public static void checkBalance(double amount) throws InsufficientFundsException {
        double balance = 100;
        if(amount > balance) {
            throw new InsufficientFundsException(amount - balance);
        }
    }
    
    //Multiple throws
    public static void processData() throws IOException, ClassNotFoundException {
        //Method may throw either exception
    }
}

//============================================================
//CUSTOM EXCEPTION DEFINITION - Checked
//============================================================
class InsufficientFundsException extends Exception {
    private double deficit;
    
    public InsufficientFundsException(double deficit) {
        super("Insufficient funds");
        this.deficit = deficit;
    }
    
    public double getDeficit() {
        return deficit;
    }
}

//============================================================
//CUSTOM EXCEPTION DEFINITION - Unchecked
//============================================================
class InvalidInputException extends RuntimeException {
    public InvalidInputException(String message) {
        super(message);
    }
}

//============================================================
//EXCEPTION HIERARCHY
//============================================================
//Throwable
//├── Error (unchecked - system errors, don't catch)
//│   └── OutOfMemoryError, StackOverflowError, etc.
//└── Exception
//    ├── RuntimeException (unchecked - programming errors)
//    │   ├── NullPointerException
//    │   ├── ArrayIndexOutOfBoundsException
//    │   ├── ArithmeticException
//    │   ├── IllegalArgumentException
//    │   ├── ClassCastException
//    │   └── NumberFormatException
//    └── Checked Exceptions (must handle or declare)
//        ├── IOException
//        ├── SQLException
//        ├── ClassNotFoundException
//        └── FileNotFoundException